toolchain/check/pattern_block_stack.h

```diff
@@ -0,0 +1,38 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_PATTERN_BLOCK_STACK_H_
+#define CARBON_TOOLCHAIN_CHECK_PATTERN_BLOCK_STACK_H_
+
+#include "common/array_stack.h"
+#include "toolchain/sem_ir/ids.h"
+
+namespace Carbon::Check {
+
+class Context;
+
+// Stack of blocks containing pattern insts.
+class PatternBlockStack {
```

Comment:
Had you reconsidered reusing InstBlockStack? There's a lot of logic there that might apply here too: avoiding allocation for empty blocks, debug logging, printing support for stack traces, and so on.